### DISCUSSION TITLE
Other way for returning multiple arguments (compared to existing tuple syntaxes)

### DISCUSSION THREAD

**Discussion by CitizenInsane** [dotnet/csharplang#5240] [Language Ideas]
**[Post edited ==> Just to put idea at the very beginning / No change otherwise]**

It may be nice that all `out` arguments appearing at the beginning of a function returning `void` could then be declared/used with similar tuple syntax as introduced in C#7.0.

For instance, declaring:

    public [double x, int y, bool[] z] = foo(string w) { ... }

Would "just" translate/stand for:

    public void foo(out double x, out int y, out bool[] z, string w) { ... }

and could be called as:

    [var x] = foo(w); // Just interested in x (eventually simply `var x = foo(w);`, i.e. with no brackets) 
    [var x, _, var z] = foo(w); // Interested in x and z but not y

That is translating/standing for;

    foo(out var x, out var yNotUsed, out var zNotUsed, w);
    foo(out var x, out var yNotUsed, out var z, w);

I know that with tuple syntax introduced in C#7.0:

    public (double x, int y, bool[] z) = foo(string w) { ... }
    
I can already achieve quite the same with:

    var x = foo(w).x;
    var z = foo(w).z;
    (var x, _, var z) = foo(w);

But:

- Writing **always** the same `[...out...] = somefunction(..in/ref..);` syntax both at declaration and usage times appears more concise
- **Far more importantly**, it would be possible this way to "return" (i.e. "output") various "deconstructed-tuples"  for the same function signature without failing into ambiguities (at least they can be solved with explicit typing):

.

    // Not possible to have different "return" types for same inputs ==> because in C# functions signature only depends on inputs
    public (double[], int[]) somefunction(...in/ref...) { .... }
    public (double[], double[]) somefunction(...in/ref...) { .... }
    
    // Yes this is already possible
    public void somefunction(out double[] o1, out int[] o2, ....in/ref...) { ... }
    public void somefunction(out double[] o1, out double[] o2, ....in/ref...) { ... }
    
    // And ambiguities (if any) could be "easily" removed using explicit typing:
    // NB: Just as it is already possible with explicit call like `somefunction(out var o1, out int[] o2, ...in/ref...);`
    [var o1, int[] o2] = somefunction(...in/ref...);
    [var o1, double[] o2] = somefunction(...in/ref...);

I'm particularly interested in this last case of "various" returned "tuples" for same input signature because I already think of cases of mine where it is extremely time consuming to evaluate all outputs (because recursive of one-another):

    // No possible optimization (!!! same signature !!)
    public (double[,] R) RadialEval(int n, double k, double r);
    public (double[,] R, double[,] dR) RadialEval(int n, double k, double r);
    public (double[,] R, double[,] dR, double[,] Rc) RadialEval(int n, double k, double r);

    // Already some possible optimization
    public void RadialEval(out double[,] R, int n, double k, double r) { radialEvalCore(out R, out double[,] dR, out double[,] Rc, n, k, r, 1 ); }
    public void RadialEval(out double[,] R, out double[,] dR, int n, double k, double r) { radialEvalCore(out R, out dR, out double[,] Rc, n, k, r, 2 ); }
    public void RadialEval(out double[,] R, out double[,] dR, out double[,] Rc, int n, double k, double r) { radialEvalCore(out R, out dR, out Rc, n, k, r, 3 ); }
    private void radialEvalCore(out double[,] R, out double[,] dR, out double[,] Rc, int n, double k, double r, int nargout)
    {
         // Init
         R = null; dR = null; Rc = null;
   
         // Compute only the minimum the user is interested in
         if (nargout >= 1) {
             R = ... extremely time consuming ...
             if (nargout >= 2) {
                 dR = ... extremely time consuming depending on R ...
                 if (nargout >= 3) {
                     Rc = ... extremely time consuming depending on R and dR ...
                 }
             }
         }    
     }

     // And then would be nice, for functions with such "out" synopsis, to benefit from:
     var R = RadialEval(n, k, r);
     [var R, var dR] = RadialEval(n, k, r);
     [var R, _, var Rc] = RadialEval(n, k, r);

I'm not hiding I'm coming from Matlab world as you may already have suspected and that I love `function [varargout] = foo(varargin)` syntax as described above because it is easy to understand and then write code. Imitation of this syntax with friendly syntax basically translating to `void foo(...out... , ...in/ref....)` would be great (and not too complex or at least not a big paradigm change -- it's just the compiler doing the painful translation automatically for us -- or with manual explicit typing to remove ambiguities if any --). 

Some limitations I can see:

- Imitating Matlab's `varargout` seems possible but fully not type-safe : `public void foo(params out object[] ouputs`, ...in/ref...) { ... }`
- Would be nice to know which output arguments are not used e.g. `(var R, _, var Rc)` but don't know how (it is not even possible in Matlab or only with some tricks).

**Comment by HaloFour** [comment]
This syntax was considered and rejected as a part of the design for tuples.  Combined with the existing tuple syntax it would make overloads much more complicated.  I also don't believe that it would enable the input overloading as you suggest given that such overload resolution is not based on the target typing of the elements but on the arity of the elements.  But that would only complicate overload resolution that much more.

**Reply by CitizenInsane** [reply]
I don't clearly see your point about overloading because syntax `void foo(out(s), ....in/ref...)` already exists and allows output overloading using different types for `out` arguments. Then calling `foo` as `(...outs...) = foo(...in/ref...)` (or eventually as `[var o1, double o2] = foo(...in/ref...)`, i.e. using bracket to distinguish from tuples and/or explicit typing for overload resolution if necessary) just appears to me as syntax sugar on how one would currently call functions like `void foo(out(s), ....in/ref...)` and let the compiler make the conversion to "canonical" syntax for us. I'm not skilled enough to see all the implications, anyway, if this has already been considered, it's already fine for me (even if rejected).
